src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs

```diff
@@ -323,13 +323,45 @@ private static void ValidateHashAlgorithm(HashAlgorithmName hashAlgorithm)
             ArgumentException.ThrowIfNullOrEmpty(hashAlgorithmName, nameof(hashAlgorithm));
 
             // MD5 intentionally left out.
-            if (hashAlgorithmName != HashAlgorithmName.SHA1.Name &&
-                hashAlgorithmName != HashAlgorithmName.SHA256.Name &&
-                hashAlgorithmName != HashAlgorithmName.SHA384.Name &&
-                hashAlgorithmName != HashAlgorithmName.SHA512.Name)
+            if (hashAlgorithmName == HashAlgorithmName.SHA1.Name ||
+                hashAlgorithmName == HashAlgorithmName.SHA256.Name ||
+                hashAlgorithmName == HashAlgorithmName.SHA384.Name ||
+                hashAlgorithmName == HashAlgorithmName.SHA512.Name)
             {
-                throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmName));
+                return;
+            }
+
+            if (hashAlgorithmName == HashAlgorithmName.SHA3_256.Name)
+            {
+                if (HMACSHA3_256.IsSupported)
+                {
+                    return;
+                }
+
+                throw new PlatformNotSupportedException();
+            }
+
+            if (hashAlgorithmName == HashAlgorithmName.SHA3_384.Name)
+            {
+                if (HMACSHA3_384.IsSupported)
```

Comment:
It's probably fine for the product code to make the same "one SHA3, all SHA3" assumption that the tests do (they'd be proven invalid at the same time, at least).  So these could be

```C#
// If one SHA-3 size is supported, all three are
if (hashAlgorithmName == HashAlgorithmName.SHA3_256.Name ||
    hashAlgorithmName == HashAlgorithmName.SHA3_384.Name ||
    hashAlgorithmName == HashAlgorithmName.SHA3_512.Name)
{
    if (HMACSHA3_256.IsSupported)
    {
        return;
    }

    throw new PlatformNotSupportedException();
}
```

Reply:
I waffled on this. To a casual reviewer it might look like a bug, but I think with a sufficient comment it makes sense.